Port processing densify algorithms to c++

nyalldawson opened the pull request:
- add new API QgsGeometry::densifyByCount and QgsGeometry::densifyByDistance
- add unit tests for above
- port processing densify algorithms to use c++ version and fix bugs in existing processing implementation- - fix spelling errors in processing alg help

But that's all just side effects of the main goal of this PR - densifying the canvas extent geometry before reprojecting for map views with different CRSes. This change makes the canvas extent indicator more accurately reprojected in views.

nyalldawson committed:
[FEATURE] New API QgsGeometry::densifyByCount

Densifies a geometry by adding a specified number of vertices
to each segment

nyalldawson committed:
Densify canvas extent geometry before reprojecting for views

Makes displaying canvas extent in extra map views nice and smooth
if canvas CRSes differ.

nyalldawson committed:
Port processing Densify Geometries algorithm to use c++ densify

Rough benchmark using 50k geometry polygon shapefile

Before: 1m25s
After:  21s

nyalldawson committed:
Fix spelling errors in processing algorithm help

nyalldawson committed:
Also port processing densify to distance to c++

- Add QgsGeometry method to densify by distance
- Fix bug in processing algorithm which resulted in duplicate
vertices and incorrectly spaced extra vertices

nirvn commented:
Ahh that's great. Proper map extent onto orthographic projection map view is great.

Would be worth doing the same thing of composer map item overview extent too.

nyalldawson committed:
Fix doxygen warning

nyalldawson committed:
[composer] Correctly draw overview frame when map item CRS differs

nyalldawson committed:
Add optimised constructors for QgsLineString

Instead of requiring all linestrings to be constructed by
first creating QgsPointSequence (requiring creation or
conversion of points to QgsPointV2), allow construction
of LineStrings directly from vectors of values (fastest!)
or lists of QgsPoint.

Likely results in speedups for lots of geometry operations,
but using the same layer as earlier tested for densify
improvements the densify operation time dropped further
from 25 seconds to 15 seconds.
